### PR TITLE
try to run cibuildwheel on jenkins

### PR DESCRIPTION
**Pull request by beniwohli**

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -137,6 +137,23 @@ pipeline {
...
+        deleteDir()
+        unstash 'source'
+        dir("${BASE_DIR}"){
+          sh script: 'pip3 install --user cibuildwheel', label: "Installing cibuildwheel"
+          sh script: 'mkdir wheelhouse', label: "creating wheelhouse"
+          sh script: 'cibuildwheel --platform linux --output-dir wheelhouse; ls -l wheelhouse'
+        }
+      }
+    }
   }
```
Comment (kuisathaverat) [inline Jenkinsfile]:
```suggestion
      stage('Release') {
        agent { label 'linux && immutable' }
        options { skipDefaultCheckout() }
        environment {
          HOME = "${env.WORKSPACE}"
          PATH = "${env.PATH}:${env.WORKSPACE}/.local/bin"
        }
        when {
          beforeAgent true
          anyOf {
            tag "v\\d+\\.\\d+\\.\\d+*"
            expression { return params.Run_As_Master_Branch }
          }
        }
        steps {
          input(message: 'Should we release a new version?', ok: 'Yes, we should.')
          deleteDir()
          unstash 'source'
          unstash('packages')
          dir("${BASE_DIR}"){
            /** TODO code to release */
          }
        }
      }
    }
```

File: Jenkinsfile [hunk 2]
```diff
@@ -137,6 +137,23 @@ pipeline {
...
+        PATH = "${env.PATH}:${env.WORKSPACE}/.local/bin"
+      }
+      steps {
+        deleteDir()
+        unstash 'source'
+        dir("${BASE_DIR}"){
+          sh script: 'pip3 install --user cibuildwheel', label: "Installing cibuildwheel"
+          sh script: 'mkdir wheelhouse', label: "creating wheelhouse"
+          sh script: 'cibuildwheel --platform linux --output-dir wheelhouse; ls -l wheelhouse'
+        }
```
Comment (kuisathaverat) [inline Jenkinsfile]:
```suggestion
          stash allowEmpty: true, name: 'packages', includes: "path_to_packages/**", useDefaultExcludes: false
        }
```

Comment (kuisathaverat) [reply Jenkinsfile]: this store temporarily the packages in Jenkins, we can also archive them in Jenkins with the artifacts step

```
artifacts: "path_to_packages/**", onlyIfSuccessful: true)
```

File: Jenkinsfile (hunk 1 above)
Comment (kuisathaverat) [reply Jenkinsfile]: this stage will be trigger when we create a tag like vNumber.Number.Number, it'd take the packages that we stash before and push them to some place

Comment (beniwohli) [reply Jenkinsfile]: Nice!

The code to push the release is relatively simple:

https://github.com/elastic/apm-agent-python/blob/b3b0ba42c260935b1d3c3a1db244c1ebc14d40ad/.travis.yml#L36-L39

(we'd also need to make the PyPI password available as environment variable)

Comment (kuisathaverat) [reply Jenkinsfile]: we can use getVaultSecret('secret') to grab the credentials something like this

```
  def jsonValue = getVaultSecret(secret: 'secret')
  wrap([$class: 'MaskPasswordsBuildWrapper', varPasswordPairs: [
    [var: 'TWINE_USER', password: jsonValue.data.user],
    [var: 'TWINE_PASSWORD', password: jsonValue.data.password],
    ]]) {
    withEnv([
      "TWINE_USER=${jsonValue.data.user}",
      "TWINE_PASSWORD=${jsonValue.data.password}"]) {
        sh(label: "Release packages", script: """
        python -m pip install twine 
        python setup.py sdist 
        python -m twine upload --skip-existing --repository-url https://upload.pypi.org/legacy/ dist/*.tar.gz 
        python -m twine upload --skip-existing --repository-url https://upload.pypi.org/legacy/ wheelhouse/*.whl
        """)
    }
  }
```

mind note: we need a `withSecretVault('secret'){ /** code block */ }` step to make things simpler

File: Jenkinsfile [hunk 3]
```diff
@@ -137,7 +137,49 @@ pipeline {
...
+            expression { return params.Run_As_Master_Branch }
+          }
+        }
+        steps {
+          input(message: 'Should we release a new version?', ok: 'Yes, we should.')
+          deleteDir()
+          unstash 'source'
+          unstash('packages')
+          dir("${BASE_DIR}"){
+            /** TODO code to release */
```
Comment (kuisathaverat) [inline Jenkinsfile]:
```suggestion
  def jsonValue = getVaultSecret(secret: 'secret')
  wrap([$class: 'MaskPasswordsBuildWrapper', varPasswordPairs: [
    [var: 'TWINE_USER', password: jsonValue.data.user],
    [var: 'TWINE_PASSWORD', password: jsonValue.data.password],
    ]]) {
    withEnv([
      "TWINE_USER=${jsonValue.data.user}",
      "TWINE_PASSWORD=${jsonValue.data.password}"]) {
        sh(label: "Release packages", script: """
        set +x
        python -m pip install twine 
        python setup.py sdist 
        python -u \${TWINE_USER} -p \${TWINE_PASSWORD} -m twine upload --skip-existing --repository-url https://upload.pypi.org/legacy/ dist/*.tar.gz 
        python -u \${TWINE_USER} -p \${TWINE_PASSWORD} -m twine upload --skip-existing --repository-url https://upload.pypi.org/legacy/ wheelhouse/*.whl
        """)
    }
  }            
```

File: Jenkinsfile [hunk 4]
```diff
@@ -137,7 +137,49 @@ pipeline {
...
+        PATH = "${env.PATH}:${env.WORKSPACE}/.local/bin"
+      }
+      steps {
+        deleteDir()
+        unstash 'source'
+        dir("${BASE_DIR}"){
+          sh script: 'pip3 install --user cibuildwheel', label: "Installing cibuildwheel"
+          sh script: 'mkdir wheelhouse', label: "creating wheelhouse"
+          sh script: 'cibuildwheel --platform linux --output-dir wheelhouse; ls -l wheelhouse'
+          stash allowEmpty: true, name: 'packages', includes: "path_to_packages/**", useDefaultExcludes: false
```
Comment (kuisathaverat) [inline Jenkinsfile]:
```suggestion
          stash allowEmpty: true, name: 'packages', includes: "wheelhouse/*.whl,dist/*.tar.gz", useDefaultExcludes: false
```

File: Jenkinsfile (hunk 3 above)
Comment (beniwohli) [reply Jenkinsfile]: A couple of things:

 * we can't install Python modules globally due to access rights. So we need to use the `--user` flag:

        python3 -m pip install --user twine

 * the `-u` /  `-p` flags need to be given to `twine upload`, not `python`

        python3 -m twine upload -u \${TWINE_USER} -p \${TWINE_PASSWORD} --skip-existing --repository-url https://upload.pypi.org/legacy/ dist/*.tar.gz

File: Jenkinsfile [hunk 5]
```diff
@@ -137,7 +137,49 @@ pipeline {
...
+        }
+        when {
+          beforeAgent true
+          anyOf {
+            tag "v\\d+\\.\\d+\\.\\d+*"
+            expression { return params.Run_As_Master_Branch }
+          }
+        }
+        steps {
+          input(message: 'Should we release a new version?', ok: 'Yes, we should.')
```
Comment (kuisathaverat) [inline Jenkinsfile]:
```suggestion
          input(message: 'Should we release a new version?', ok: 'Yes, we should.')
          input(
            message: 'Should we release a new version?', 
            ok: 'Yes, we should.', 
            parameters: [
              choice(
                choices: [
                  'https://test.pypi.org/legacy/', 
                  'https://upload.pypi.org/legacy/'
                 ],
                 description: 'PyPi repository URL', 
                 name: 'REPO_URL')
               ])
```

File: Jenkinsfile (hunk 3 above)
Comment (kuisathaverat) [reply Jenkinsfile]: I've made the change

**Pull request closed without merging**